### PR TITLE
Web Inspector: Show parent style rules for nested style rules in Styles sidebar

### PR DESCRIPTION
**Pull request by patrickangle**
#### b5cf192b43bc91e0a18cf1399fd5f64d152e969c
<pre>
Web Inspector: Show parent style rules for nested style rules in Styles sidebar
<a href="https://bugs.webkit.org/show_bug.cgi?id=250088">https://bugs.webkit.org/show_bug.cgi?id=250088</a>
rdar://100522930

Reviewed by Antti Koivisto.

Add basic support for viewing and editing parent style rule selectors for nested style rules. This uses the same
mechanism we already use for showing parent `@rule`s, including support for editing.

* LayoutTests/inspector/css/getMatchesStylesForNodeNestingStyleGrouping-expected.txt: Added.
* LayoutTests/inspector/css/getMatchesStylesForNodeNestingStyleGrouping.html: Added.

* LayoutTests/inspector/css/setGroupingHeaderText-expected.txt:
* LayoutTests/inspector/css/setGroupingHeaderText.html:
- Add test cases for modifying a parent Style rule&apos;s selector via a Grouping on a child nested rule.

* Source/JavaScriptCore/inspector/protocol/CSS.json:
- Add new &quot;StyleRule&quot; grouping type to represent parent style rules.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::CSSStyleRule):
(WebCore::CSSStyleRule::length const):
(WebCore::CSSStyleRule::item const):
(WebCore::CSSStyleRule::cssRules const):
* Source/WebCore/css/CSSStyleRule.h:
- Add support for getting the nested rules inside a CSSStyleRule.

* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRuleBase::createCSSOMWrapper const):
* Source/WebCore/css/StyleRule.h:
- Support creating a wrapper with a CSSStyleRule as a parent.

* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::parseSelector):
* Source/WebCore/css/parser/CSSParser.h:
- Allow callers to enable nesting syntax for parsing a selector, which is done to verify a selector is valid by Web
Inspector when editing a nested rule&apos;s selector.

* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::flatteningStrategyForStyleRuleType):
- Style rules can now contain rules, so get rid of the `Commit` strategy and use the existing `CommitSelfThenChildren`
for style rules instead.

(WebCore::isValidRuleHeaderText):
- Pass through the nesting mode for parsing selectors.

(WebCore::protocolGroupingTypeForStyleRuleType):
(WebCore::flattenSourceData):
- Get rid of `Commit` strategy.

(WebCore::StyleSheetHandler::startRuleHeader):
- It is no longer invalid to encounter the start of a style rule before encountering the end of the previous style rule.

(WebCore::asCSSRuleList):

(WebCore::InspectorStyleSheet::buildArrayForGroupings):
- Start with the parent of the passed CSSRule, otherwise every style rule will include itself as a grouping.

(WebCore::isNestedContext):
(WebCore::InspectorStyleSheet::setRuleHeaderText):
- Nested rules should allow relevant syntax for selectors.

(WebCore::InspectorStyleSheet::collectFlatRules):

* Source/WebCore/style/InspectorCSSOMWrappers.cpp:
(WebCore::Style::InspectorCSSOMWrappers::collect):
- Eagerly create CSSOM wrappers for nested rules.

* Source/WebInspectorUI/UserInterface/Models/CSSGrouping.js:
(WI.CSSGrouping.prototype.get isStyle):
(WI.CSSGrouping.prototype.get prefix):
(WI.CSSGrouping):
- Nested rules don&apos;t have a prefix like `@rule`s do, so provide a null prefix.

* Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._renderGroupings):
- Support groupings without a prefix by not prepending the prefix.

Canonical link: <a href="https://commits.webkit.org/258555@main">https://commits.webkit.org/258555@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/e34f11836365ee897be96ddd9a7f75bb77f77920

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102325 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11463 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35391 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111634 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/171809 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106301 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12463 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2382 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94654 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109344 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108106 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9534 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/92802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37279 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91402 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24287 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/79026 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/92634 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4976 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25710 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/88890 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/2640 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5114 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2151 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29542 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11144 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45203 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/91814 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5878 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6867 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20540 "Passed tests") | 
| | | | |

### CODE REVIEW

**Review by anttijk — APPROVED** [review]

File: Source/WebCore/inspector/InspectorStyleSheet.cpp [hunk 1]
```diff
@@ -1112,7 +1122,7 @@ ExceptionOr<CSSStyleRule*> InspectorStyleSheet::addRule(const String& selector)
     if (!m_pageStyleSheet)
         return Exception { NotSupportedError };
 
-    if (!isValidRuleHeaderText(selector, StyleRuleType::Style, m_pageStyleSheet->ownerDocument()))
+    if (!isValidRuleHeaderText(selector, StyleRuleType::Style, m_pageStyleSheet->ownerDocument(), CSSSelectorParser::IsNestedContext::No))
```
Comment (mdubet) [inline Source/WebCore/inspector/InspectorStyleSheet.cpp:1129]: > When used in the selector of a [nested style rule](https://drafts.csswg.org/css-nesting/#nested-style-rule), the [nesting selector](https://drafts.csswg.org/css-nesting/#nesting-selector) represents the elements matched by the parent rule. When used in any other context, it represents nothing. (That is, it’s valid, but matches no elements.)

Actually the nesting selector & is allowed everywhere (even inside top-level rules), so maybe we don't need this parameter.

File: Source/WebCore/css/CSSStyleRule.cpp [hunk 2]
```diff
@@ -45,6 +46,7 @@ CSSStyleRule::CSSStyleRule(StyleRule& styleRule, CSSStyleSheet* parent)
     : CSSRule(parent)
     , m_styleRule(styleRule)
     , m_styleMap(DeclaredStylePropertyMap::create(*this))
+    , m_childRuleCSSOMWrappers(styleRule.nestedRules().size())
```
Comment (mdubet) [inline Source/WebCore/css/CSSStyleRule.cpp:49]: Should we populate this directly instead of doing it in the `item()` function ?  
Maybe this would help to get rid of the `mutable` flag also ?

File: Source/WebCore/css/CSSStyleRule.h [hunk 3]
```diff
@@ -46,6 +47,10 @@ class CSSStyleRule final : public CSSRule, public CanMakeWeakPtr<CSSStyleRule> {
     // FIXME: Not CSSOM. Remove.
     StyleRule& styleRule() const { return m_styleRule.get(); }
 
+    CSSRuleList& cssRules() const;
```
Comment (mdubet) [inline Source/WebCore/css/CSSStyleRule.h:50]: https://drafts.csswg.org/css-nesting/#issue-e535b5ff

This is a high probability that the spec will move this to CSSStyleDeclaration, but we can start with this until the spec is finalised.

File: Source/WebCore/inspector/InspectorStyleSheet.cpp [hunk 4]
```diff
@@ -296,10 +292,6 @@ class StyleSheetHandler : public CSSParserObserver {
 
 void StyleSheetHandler::startRuleHeader(StyleRuleType type, unsigned offset)
 {
-    // Pop off data for a previous invalid rule.
```
Comment (mdubet) [inline Source/WebCore/inspector/InspectorStyleSheet.cpp]: Is this related to this patch?

File: Source/WebCore/inspector/InspectorStyleSheet.cpp (hunk 1 above)
Comment (mdubet) [reply Source/WebCore/inspector/InspectorStyleSheet.cpp:1129]: But the relative selector syntax is not allowed everywhere though, so maybe we do need this here (depending of the goal of the inspector)

Comment (patrickangle) [reply Source/WebCore/inspector/InspectorStyleSheet.cpp:1129]: On the line you commented on here, we will always be creating a new top-level style rule, so the relative syntax wouldn’t be allowed if I understand correctly. Above where we allow editing the selector, we dynamically determine if nesting-only syntax should be allowed. Our goal is to ensure that an edit to a selector maintains the validity of the overall style sheet (even if the rule no longer applies to the element after the edit). We don’t want developers to be able to make an edit they can’t fix later because the style is no longer syntactically valid, preventing us from parsing it and potentially other styles.

File: Source/WebCore/css/CSSStyleRule.h (hunk 3 above)
Comment (patrickangle) [reply Source/WebCore/css/CSSStyleRule.h:50]: I purposely didn’t web-expose this since it’s still in flux. Just needed it somewhere to start, and for now matching the spec and CSSGroupingRule seemed like a safe choice.

File: Source/WebCore/inspector/InspectorStyleSheet.cpp (hunk 4 above)
Comment (patrickangle) [reply Source/WebCore/inspector/InspectorStyleSheet.cpp]: Yes. We can no longer pop off the rule data here because that would pop off the parent data for a nested rule, but we haven’t reached the end of the parent rule’s body yet.

File: Source/WebCore/css/CSSStyleRule.cpp (hunk 2 above)
Comment (patrickangle) [reply Source/WebCore/css/CSSStyleRule.cpp:49]: For consistency I was trying to closely match what CSSGroupingRule does to prevent there being any discrepancy in how these lists of rules are created.

File: Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js [hunk 5]
```diff
@@ -330,7 +330,10 @@ WI.SpreadsheetCSSStyleDeclarationSection = class SpreadsheetCSSStyleDeclarationS
             let grouping = groupings[i];
             let groupingTypeElement = this._groupingsContainerElement.appendChild(document.createElement("div"));
             groupingTypeElement.classList.add("grouping");
-            groupingTypeElement.textContent = grouping.prefix + " ";
+
+            let prefix = grouping.prefix;
+            if (prefix)
+                groupingTypeElement.textContent = prefix + " ";
```
Comment (rcaliman-apple) [inline Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:336]: Will need to also handle the `null` prefix in `CSSStyleDeclaration.generateFormattedText()`.

https://github.com/WebKit/WebKit/blob/60f4fa76729bafa5b2594c42d3eb0996ebe2838e/Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js#L539

That's invoked when using the "Copy Rule" menu item when clicking the `{}` icon next to a CSS rule in the Styles panel.

Screenshot 2023-01-05 at 6 15 35 PM

**Review by rcaliman-apple — APPROVED** [review]
The front-end parts build elegantly on top of your prior work. Good job 👍

Still pondering how better to display the parent rules to communicate what they are, but this work stands on its own.

Found a couple of other issues with groupings, but I'll open separate bugs:
- tab navigation forwards from a previous rule skips the next one's groupings; backwards navigation works fine.
Update: logged [Bug 250134](https://bugs.webkit.org/show_bug.cgi?id=250134)

- it would be nice to include the grouping prefix when filtering in the Styles panel 
Filtering for the string `media` won't match `@media (width: 1200px)`, but filtering for anything in the condition text will.
Update: logged [Bug 250136](https://bugs.webkit.org/show_bug.cgi?id=250136)

File: Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js (hunk 5 above)
Comment (patrickangle) [reply Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:336]: Good catch!

File: Source/WebCore/inspector/InspectorStyleSheet.cpp (hunk 4 above)
Comment (patrickangle) [reply Source/WebCore/inspector/InspectorStyleSheet.cpp]: But also after looking further at what is changing in this patch, removing this isn't necessary.
